[🐛 Bug]: Hub and Node BiDi WebSocket connection fails in Docker / VPC setup

### Description

I am experiencing an issue where Selenium Grid Hub and Node fail to establish a BiDi WebSocket connection when deployed in Docker containers across a VPC. The same setup works fine when Hub and Node are running on local machines (no containers, no VPC).

### Reproducible Code

```shell
1.Start Selenium Hub in a Docker container in VPC:
java -jar selenium-server-4.38.0.jar hub --port 4444
2.Start Selenium Node in a Docker container in the same VPC:
java -jar selenium-server-4.38.0.jar \
    node \
    --hub http://<hub-private-ip>:4444 \
    --port 9000 \
    --host <node-private-ip> \
    --max-sessions 4 \
    --enable-cdp true \
    --log-level INFO
3.Create a session from client → Hub → Node (BiDi session).



Client Code:


import java.net.URL;
import java.util.Map;

import org.openqa.selenium.Capabilities;
import org.openqa.selenium.bidi.browsingcontext.BrowsingContext;
import org.openqa.selenium.bidi.browsingcontext.NavigationResult;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.remote.Augmenter;
import org.openqa.selenium.remote.RemoteWebDriver;

public class RemoteFirefoxBiDiExample {

    public static void main(String[] args) throws Exception {

        FirefoxOptions options = new FirefoxOptions();
        options.setCapability("webSocketUrl", true);
        options.addArguments("--headless");
        options.addArguments("--no-sandbox"); 
        options.addArguments("--disable-dev-shm-usage");
//        options.setCapability("se:bidiEnabled", true);

        RemoteWebDriver 
        driver = new RemoteWebDriver(
            new URL("http://IP:4444/wd/hub"),
            options
        );
 
        // Print WebSocket URL
        Capabilities caps = driver.getCapabilities();
        Object wsUrl = caps.getCapability("webSocketUrl");
        System.out.println("WebSocket URL: " + wsUrl);
        System.out.println(driver.getCapabilities().asMap());


        driver = (RemoteWebDriver) new Augmenter().augment(driver);
        System.out.println(driver instanceof org.openqa.selenium.bidi.HasBiDi);

        // Create BiDi browsing context
      BrowsingContext context = new BrowsingContext(driver, driver.getWindowHandle());
       // Navigate using BiDi
       NavigationResult result =  context.navigate("https://example.com");

       System.out.println("Navigation ID: " + result.getNavigationId());
        System.out.println("URL: " + result.getUrl());

      driver.quit();
    }
}
```

### Debugging Logs

```logs
Hub Logs:

[2026-02-13 16:10:30]  [stdout] 1024
  [2026-02-13 16:10:30]  [stdout] Starting HUB
  [2026-02-13 16:10:31]  [INFO] InactivityTimeRefresher task has been started at 1,770,979,231,825
  [2026-02-13 16:10:31]  [INFO] SessionQueueWatcher task has been started 1,770,979,231,831
  [2026-02-13 16:10:31]  [INFO] DiskUsageWatcher task has been started for Standalone at 1,770,979,231,832
  [2026-02-13 16:10:32]  [INFO] Started Selenium Hub 4.38.0 (revision 6b412e825c): http://10.71.67.180:30001
  [2026-02-13 16:10:37]  [INFO] Node creation started 1,770,979,237,192
  [2026-02-13 16:10:39]  [INFO] Binding additional locator mechanisms: relative
  [2026-02-13 16:10:39]  [INFO] Switching Node 39e338af-64d5-48f6-bb5b-3bbe078b639d (uri: http://10.0.0.5:9000) from DOWN to UP
  [2026-02-13 16:10:39]  [INFO] Added node 39e338af-64d5-48f6-bb5b-3bbe078b639d at http://10.0.0.5:9000. Health check every 120s
  [2026-02-13 16:10:39]  [INFO] Session request received by the Distributor: \n [Capabilities {acceptInsecureCerts: true, browserName: firefox, moz:firefoxOptions: {args: [--headless, --no-sandbox, --disable-dev-shm-usage], prefs: {remote.active-protocols: 1}}, webSocketUrl: true}]
  [2026-02-13 16:10:43]  [INFO] Added session to local Session Map, Id: 336b8e95-863b-4d5a-a3e5-a65f93e81c50, Node: http://10.0.0.5:9000
  [2026-02-13 16:10:43]  [INFO] Session created by the Distributor. Id: 336b8e95-863b-4d5a-a3e5-a65f93e81c50 \n Caps: Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 140.0, moz:accessibilityChecks: false, moz:buildID: 20250616215311, moz:firefoxOptions: {args: [--headless, --no-sandbox, --disable-dev-shm-usage], prefs: {remote.active-protocols: 1}}, moz:geckodriverVersion: 0.36.0, moz:headless: true, moz:platformVersion: 4.19.36, moz:processID: 674, moz:profile: /tmp/rust_mozprofilew6zHvM, moz:shutdownTimeout: 60000, moz:webdriverClick: true, moz:windowless: false, pageLoadStrategy: normal, platformName: linux, proxy: {}, se:cdpEnabled: false, se:downloadsEnabled: true, se:gridWebSocketUrl: ws://127.0.0.1:39743/sessio..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, userAgent: Mozilla/5.0 (X11; Linux x86..., webSocketUrl: ws://10.71.67.180:30001/ses...}
  [2026-02-13 16:10:43]  [INFO] WebSocket request received for URI: /session/336b8e95-863b-4d5a-a3e5-a65f93e81c50/se/bidi
  [2026-02-13 16:10:43]  [INFO] Extracted sessionId: 336b8e95-863b-4d5a-a3e5-a65f93e81c50
  [2026-02-13 16:10:43]  [INFO] Resolved session URI (node location): http://10.0.0.5:9000
  [2026-02-13 16:10:43]  [INFO] Attempting to open WebSocket to upstream. Base URI: http://10.0.0.5:9000 , Request URI: /session/336b8e95-863b-4d5a-a3e5-a65f93e81c50/se/bidi
  [2026-02-13 16:10:43]  [SEVERE] Connecting to upstream websocket failed. sessionId=336b8e95-863b-4d5a-a3e5-a65f93e81c50, sessionUri=http://10.0.0.5:9000, requestUri=/session/336b8e95-863b-4d5a-a3e5-a65f93e81c50/se/bidiorg.openqa.selenium.remote.http.ConnectionFailedException: JdkWebSocket initial request execution error\nBuild info: version: '4.38.0', revision: '6b412e825c'\nSystem info: os.name: 'Linux', os.arch: 'amd64', os.version: '4.19.36', java.version: '11.0.25'\nDriver info: driver.version: unknown\n\tat org.openqa.selenium.remote.http.jdk.JdkHttpClient.openSocket(JdkHttpClient.java:253)\n\tat org.openqa.selenium.remote.tracing.TracedHttpClient.openSocket(TracedHttpClient.java:45)\n\tat org.openqa.selenium.grid.router.ProxyWebsocketsIntoGrid.apply(ProxyWebsocketsIntoGrid.java:80)\n\tat org.openqa.selenium.grid.router.ProxyWebsocketsIntoGrid.apply(ProxyWebsocketsIntoGrid.java:43)\n\tat org.openqa.selenium.netty.server.WebSocketUpgradeHandler.handleHttpRequest(WebSocketUpgradeHandler.java:137)\n\tat org.openqa.selenium.netty.server.WebSocketUpgradeHandler.channelRead(WebSocketUpgradeHandler.java:100)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:356)\n\tat io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)\n\tat io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.onHttpRequestChannelRead(WebSocketServerExtensionHandler.java:158)\n\tat io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:82)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:354)\n\tat io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)\n\tat io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:354)\n\tat io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:434)\n\tat io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)\n\tat io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:249)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:354)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1429)\n\tat io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)\n\tat io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:168)\n\tat io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.handle(AbstractNioChannel.java:445)\n\tat io.netty.channel.nio.NioIoHandler$DefaultNioRegistration.handle(NioIoHandler.java:388)\n\tat io.netty.channel.nio.NioIoHandler.processSelectedKey(NioIoHandler.java:596)\n\tat io.netty.channel.nio.NioIoHandler.processSelectedKeysOptimized(NioIoHandler.java:571)\n\tat io.netty.channel.nio.NioIoHandler.processSelectedKeys(NioIoHandler.java:512)\n\tat io.netty.channel.nio.NioIoHandler.run(NioIoHandler.java:484)\n\tat io.netty.channel.SingleThreadIoEventLoop.runIo(SingleThreadIoEventLoop.java:225)\n\tat io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:196)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1193)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.net.http.WebSocketHandshakeException\n\tat java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.resultFrom(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)\n\t... 1 more\nCaused by: jdk.internal.net.http.websocket.CheckFailedException: Unexpected HTTP response status code 400\n\tat java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.checkFailed(Unknown Source)\n\tat java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.handleResponse(Unknown Source)\n\t... 7 more\n
  [2026-02-13 16:10:43]  [WARNING] Connecting to upstream websocket failedorg.openqa.selenium.remote.http.ConnectionFailedException: JdkWebSocket initial request execution error\nBuild info: version: '4.38.0', revision: '6b412e825c'\nSystem info: os.name: 'Linux', os.arch: 'amd64', os.version: '4.19.36', java.version: '11.0.25'\nDriver info: driver.version: unknown\n\tat org.openqa.selenium.remote.http.jdk.JdkHttpClient.openSocket(JdkHttpClient.java:253)\n\tat org.openqa.selenium.remote.tracing.TracedHttpClient.openSocket(TracedHttpClient.java:45)\n\tat org.openqa.selenium.grid.router.ProxyWebsocketsIntoGrid.apply(ProxyWebsocketsIntoGrid.java:80)\n\tat org.openqa.selenium.grid.router.ProxyWebsocketsIntoGrid.apply(ProxyWebsocketsIntoGrid.java:43)\n\tat org.openqa.selenium.netty.server.WebSocketUpgradeHandler.handleHttpRequest(WebSocketUpgradeHandler.java:137)\n\tat org.openqa.selenium.netty.server.WebSocketUpgradeHandler.channelRead(WebSocketUpgradeHandler.java:100)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:356)\n\tat io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)\n\tat io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.onHttpRequestChannelRead(WebSocketServerExtensionHandler.java:158)\n\tat io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:82)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:354)\n\tat io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)\n\tat io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:354)\n\tat io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:434)\n\tat io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)\n\tat io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:249)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:354)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1429)\n\tat io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)\n\tat io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:168)\n\tat io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.handle(AbstractNioChannel.java:445)\n\tat io.netty.channel.nio.NioIoHandler$DefaultNioRegistration.handle(NioIoHandler.java:388)\n\tat io.netty.channel.nio.NioIoHandler.processSelectedKey(NioIoHandler.java:596)\n\tat io.netty.channel.nio.NioIoHandler.processSelectedKeysOptimized(NioIoHandler.java:571)\n\tat io.netty.channel.nio.NioIoHandler.processSelectedKeys(NioIoHandler.java:512)\n\tat io.netty.channel.nio.NioIoHandler.run(NioIoHandler.java:484)\n\tat io.netty.channel.SingleThreadIoEventLoop.runIo(SingleThreadIoEventLoop.java:225)\n\tat io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:196)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1193)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.net.http.WebSocketHandshakeException\n\tat java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.resultFrom(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)\n\t... 1 more\nCaused by: jdk.internal.net.http.websocket.CheckFailedException: Unexpected HTTP response status code 400\n\tat java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.checkFailed(Unknown Source)\n\tat java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.handleResponse(Unknown Source)\n\t... 7 more\n
  [2026-02-13 16:10:57]  [INFO] Switching Node 39e338af-64d5-48f6-bb5b-3bbe078b639d (uri: http://10.0.0.5:9000) from UP to DRAINING
  [2026-02-13 16:10:57]  [INFO] Node 39e338af-64d5-48f6-bb5b-3bbe078b639d removed and all resources cleaned up
  [2026-02-13 16:10:59]  [INFO] Binding additional locator mechanisms: relative
  [2026-02-13 16:11:00]  [INFO] Switching Node c35971bb-5435-4889-ba4a-a3df82be4197 (uri: http://10.0.0.6:9000) from DOWN to UP
  [2026-02-13 16:11:00]  [INFO] Added node c35971bb-5435-4889-ba4a-a3df82be4197 at http://10.0.0.6:9000. Health check every 120s



Node Logs:


  [2026-02-13 16:10:37]  [stdout] 1024
  [2026-02-13 16:10:37]  [stdout] Starting NODE
  [2026-02-13 16:10:37]  [stdout] Memory monitor is disabled
  [2026-02-13 16:10:37]  [stdout] Detected node IP: 10.0.0.5
  [2026-02-13 16:10:37]  [stdout] Starting Selenium Node with command:
  [2026-02-13 16:10:37]  [stdout] /var/lang/bin/java   -Dwebdriver.chrome.driver=/usr/bin/chromedriver   -Dwebdriver.gecko.driver=/usr/bin/geckodriver   -jar /var/runtime/selenium-server-4.38.0-debug.jar   --ext /var/runtime/selenium-runtime-1.0.0.jar:/var/runtime/json-20231013.jar:/var/runtime/catalyst-java-runtime-logger-1.0.0.jar   node   --port 9000   --node-implementation com.browser360.selenium.node.CustomLocalNode   --hub http://10.0.0.4:9000   --max-sessions 4   --override-max-sessions true   --registration-secret 29d33219b79881e498d2957b26aa2628   --session-timeout 300    --grid-url http://10.71.67.180:30001   --enable-managed-downloads true   --tracing false   --enable-cdp true   --log-level INFO   --host 10.0.0.5
  [2026-02-13 16:10:38]  [stdout] 16:10:38.050 INFO [LoggingOptions.configureLogEncoding] - Using the system default encoding
  [2026-02-13 16:10:38]  [stdout] 16:10:38.053 INFO [LoggingOptions.getTracer] - Using null tracer
  [2026-02-13 16:10:38]  [stdout] 16:10:38.095 INFO [UnboundZmqEventBus.<init>] - Connecting to tcp://10.0.0.4:4442 and tcp://10.0.0.4:4443
  [2026-02-13 16:10:38]  [stdout] 16:10:38.131 INFO [ZmqUtils.configureHeartbeat] - ZMQ SUB socket heartbeat configured: interval=60s, timeout=180s, ttl=360s
  [2026-02-13 16:10:38]  [stdout] 16:10:38.142 INFO [ZmqUtils.configureHeartbeat] - ZMQ PUB socket heartbeat configured: interval=60s, timeout=180s, ttl=360s
  [2026-02-13 16:10:38]  [stdout] 16:10:38.144 INFO [UnboundZmqEventBus.<init>] - Sockets created
  [2026-02-13 16:10:39]  [stdout] 16:10:39.146 INFO [UnboundZmqEventBus.<init>] - Event bus ready
  [2026-02-13 16:10:39]  [stdout] 16:10:39.211 INFO [NodeServer.createHandlers] - Reporting self as: http://10.0.0.5:9000
  [2026-02-13 16:10:39]  [stdout] 16:10:39.214 INFO [CustomLocalNode.<clinit>] - Log level for Node: INFO
  [2026-02-13 16:10:39]  [INFO] Using null tracer
  [2026-02-13 16:10:39]  [INFO] Detected 4 available processors
  [2026-02-13 16:10:39]  [WARNING] Overriding max recommended number of 4 concurrent sessions. Session stability and reliability might suffer!
  [2026-02-13 16:10:39]  [WARNING] One browser session is recommended per available processor. Safari is always limited to 1 session per host.
  [2026-02-13 16:10:39]  [WARNING] Overriding this value for Internet Explorer is not recommended. Issues related to parallel testing with Internet Explored won't be accepted.
  [2026-02-13 16:10:39]  [WARNING] Double check if enabling 'override-max-sessions' is really needed
  [2026-02-13 16:10:39]  [INFO] Looking for existing drivers on the PATH.
  [2026-02-13 16:10:39]  [INFO] Add '--selenium-manager true' to the startup command to setup drivers automatically.
  [2026-02-13 16:10:39]  [WARNING] Unable to discover proper msedgedriver version in offline mode
  [2026-02-13 16:10:39]  [INFO] Adding Firefox for {\
  [2026-02-13 16:10:39]  [INFO] Adding Chrome for {\
  [2026-02-13 16:10:39]  [INFO] Adding Edge for {\
  [2026-02-13 16:10:39]  [INFO] Binding additional locator mechanisms: relative
  [2026-02-13 16:10:39]  [INFO] Using null tracer
  [2026-02-13 16:10:39]  [INFO] Binding additional locator mechanisms: relative
  [2026-02-13 16:10:39]  [INFO] RunningSessionWatcher task started at 1,770,979,239,451
  [2026-02-13 16:10:39]  [INFO] DiskUsageWatcher task has been started for Node at 1,770,979,239,451
  [2026-02-13 16:10:39]  [INFO] Starting registration process for Node http://10.0.0.5:9000
  [2026-02-13 16:10:39]  [INFO] Started Selenium node 4.38.0 (revision 6b412e825c): http://10.0.0.5:9000
  [2026-02-13 16:10:39]  [INFO] Sending registration event...
  [2026-02-13 16:10:43]  [INFO] Session created by the Node. Id: 336b8e95-863b-4d5a-a3e5-a65f93e81c50, Caps: Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 140.0, moz:accessibilityChecks: false, moz:buildID: 20250616215311, moz:firefoxOptions: {args: [--headless, --no-sandbox, --disable-dev-shm-usage], prefs: {remote.active-protocols: 1}}, moz:geckodriverVersion: 0.36.0, moz:headless: true, moz:platformVersion: 4.19.36, moz:processID: 674, moz:profile: /tmp/rust_mozprofilew6zHvM, moz:shutdownTimeout: 60000, moz:webdriverClick: true, moz:windowless: false, pageLoadStrategy: normal, platformName: linux, proxy: Proxy(), se:cdpEnabled: false, se:downloadsEnabled: true, se:gridWebSocketUrl: ws://127.0.0.1:39743/sessio..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, userAgent: Mozilla/5.0 (X11; Linux x86..., webSocketUrl: ws://10.71.67.180:30001/ses...}
  [2026-02-13 16:10:43]  [INFO] Session successfully created with ID: 336b8e95-863b-4d5a-a3e5-a65f93e81c50
  [2026-02-13 16:10:43]  [INFO] Scheduler API call completed for session ID: 336b8e95-863b-4d5a-a3e5-a65f93e81c50
  [2026-02-13 16:10:43]  [WARNING] Too many websocket connections initiated by 336b8e95-863b-4d5a-a3e5-a65f93e81c50
  [2026-02-13 16:10:49]  [INFO] Sending registration event...
  [2026-02-13 16:10:57]  [INFO] Trying to stop all running sessions before shutting down...
```

--- a/java/src/org/openqa/selenium/grid/node/ProxyNodeWebsockets.java
+++ b/java/src/org/openqa/selenium/grid/node/ProxyNodeWebsockets.java
@@ -102,34 +102,44 @@ public class ProxyNodeWebsockets
       return Optional.empty();
     }
 
-    Session session = node.getSession(id);
-    Capabilities caps = session.getCapabilities();
-    LOG.fine("Scanning for endpoint: " + caps);
+    try {
+      Session session = node.getSession(id);
+      Capabilities caps = session.getCapabilities();
+      LOG.fine("Scanning for endpoint: " + caps);
 
-    // Used by the ForwardingListener to notify the node that the session is still active
-    Consumer<SessionId> sessionConsumer = node::isSessionOwner;
+      // Used by the ForwardingListener to notify the node that the session is still active
+      Consumer<SessionId> sessionConsumer = node::isSessionOwner;
 
-    if (bidiMatch != null) {
-      return findBiDiEndpoint(downstream, caps, sessionConsumer, id);
-    }
+      Optional<Consumer<Message>> endpoint;
+      if (bidiMatch != null) {
+        endpoint = findBiDiEndpoint(downstream, caps, sessionConsumer, id);
+      } else if (vncMatch != null) {
+        // Passing a fake consumer to the ForwardingListener to avoid sending a session notification
+        // when VNC is used.
+        sessionConsumer = fakeConsumer -> {};
+        endpoint = findVncEndpoint(downstream, caps, sessionConsumer, id);
+      } else if (fwdMatch != null) {
+        // This match happens when a user wants to do CDP over Dynamic Grid
+        LOG.info("Matched endpoint where CDP connection is being forwarded");
+        endpoint = findCdpEndpoint(downstream, caps, sessionConsumer, id);
+      } else if (caps.getCapabilityNames().contains("se:forwardCdp")) {
+        LOG.info("Found endpoint where CDP connection needs to be forwarded");
+        endpoint = findForwardCdpEndpoint(downstream, caps, sessionConsumer, id);
+      } else {
+        endpoint = findCdpEndpoint(downstream, caps, sessionConsumer, id);
+      }
 
-    if (vncMatch != null) {
-      // Passing a fake consumer to the ForwardingListener to avoid sending a session notification
-      // when VNC is used.
-      sessionConsumer = fakeConsumer -> {};
-      return findVncEndpoint(downstream, caps, sessionConsumer, id);
-    }
+      // If no endpoint could be established the connection slot must be released;
+      if (endpoint.isEmpty()) {
+        node.releaseConnection(id);
+      }
 
-    // This match happens when a user wants to do CDP over Dynamic Grid
-    if (fwdMatch != null) {
-      LOG.info("Matched endpoint where CDP connection is being forwarded");
-      return findCdpEndpoint(downstream, caps, sessionConsumer, id);
+      return endpoint;
+    } catch (Exception e) {
+      node.releaseConnection(id);
+      LOG.log(Level.WARNING, "Failed to establish WebSocket endpoint for session " + id, e);
+      return Optional.empty();
     }
-    if (caps.getCapabilityNames().contains("se:forwardCdp")) {
-      LOG.info("Found endpoint where CDP connection needs to be forwarded");
-      return findForwardCdpEndpoint(downstream, caps, sessionConsumer, id);
-    }
-    return findCdpEndpoint(downstream, caps, sessionConsumer, id);
   }
 
   private Optional<Consumer<Message>> findCdpEndpoint(
@@ -213,6 +223,10 @@ public class ProxyNodeWebsockets
       Consumer<SessionId> sessionConsumer,
       SessionId sessionId) {
     String vncLocalAddress = (String) caps.getCapability("se:vncLocalAddress");
+    if (vncLocalAddress == null) {
+      LOG.warning("No VNC endpoint address in capabilities");
+      return Optional.empty();
+    }
     Optional<URI> vncUri;
     try {
       vncUri = Optional.of(new URI(vncLocalAddress));
@@ -264,7 +278,6 @@ public class ProxyNodeWebsockets
       };
     } catch (Exception e) {
       LOG.log(Level.WARNING, "Connecting to upstream websocket failed", e);
-      node.releaseConnection(sessionId);
       client.close();
       throw e;
     }
@@ -313,6 +326,9 @@ public class ProxyNodeWebsockets
     @Override
     public void onError(Throwable cause) {
       LOG.log(Level.WARNING, "Error proxying websocket command", cause);
+      if (connectionReleased.compareAndSet(false, true)) {
+        node.releaseConnection(sessionId);
+      }
     }
   }
 }

--- a/java/src/org/openqa/selenium/grid/node/ProxyNodeWebsockets.java
+++ b/java/src/org/openqa/selenium/grid/node/ProxyNodeWebsockets.java
@@ -223,7 +223,7 @@ public class ProxyNodeWebsockets
       Consumer<SessionId> sessionConsumer,
       SessionId sessionId) {
     String vncLocalAddress = (String) caps.getCapability("se:vncLocalAddress");
-    if (vncLocalAddress == null) {
+    if (vncLocalAddress == null || vncLocalAddress.trim().isEmpty()) {
       LOG.warning("No VNC endpoint address in capabilities");
       return Optional.empty();
     }

--- a/java/test/org/openqa/selenium/grid/node/ProxyNodeWebsocketsTest.java
+++ b/java/test/org/openqa/selenium/grid/node/ProxyNodeWebsocketsTest.java
@@ -1,0 +1,406 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.grid.node;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.net.URI;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.ImmutableCapabilities;
+import org.openqa.selenium.NoSuchSessionException;
+import org.openqa.selenium.WebDriverException;
+import org.openqa.selenium.grid.data.CreateSessionRequest;
+import org.openqa.selenium.grid.data.CreateSessionResponse;
+import org.openqa.selenium.grid.data.NodeId;
+import org.openqa.selenium.grid.data.NodeStatus;
+import org.openqa.selenium.grid.data.Session;
+import org.openqa.selenium.grid.security.Secret;
+import org.openqa.selenium.internal.Either;
+import org.openqa.selenium.remote.SessionId;
+import org.openqa.selenium.remote.http.HttpClient;
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+import org.openqa.selenium.remote.http.Message;
+import org.openqa.selenium.remote.http.WebSocket;
+import org.openqa.selenium.remote.tracing.DefaultTestTracer;
+
+class ProxyNodeWebsocketsTest {
+
+  private static final Secret SECRET = new Secret("test");
+
+  /**
+   * Regression test for counter leak: if tryAcquireConnection succeeds but no WebSocket endpoint
+   * can be established, releaseConnection must be called so the slot is returned to the counter.
+   * Without the fix, the counter climbs on every failed attempt until it reaches the configured
+   * limit, blocking all future connections regardless of how high the limit is set.
+   */
+  @Test
+  void shouldReleaseConnectionSlotWhenNoEndpointCanBeEstablished() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger acquireCount = new AtomicInteger();
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    // Capabilities with no CDP endpoint caps (no goog:chromeOptions / ms:edgeOptions) so that
+    // findCdpEndpoint iterates over all known caps, finds nothing, and returns Optional.empty().
+    Session session =
+        new Session(
+            sessionId,
+            URI.create("http://localhost:4444"),
+            new ImmutableCapabilities(),
+            new ImmutableCapabilities(),
+            Instant.now());
+
+    Node node = new CountingStubNode(sessionId, session, acquireCount, releaseCount);
+
+    // clientFactory is never reached because endpoint discovery fails before createWsEndPoint.
+    ProxyNodeWebsockets proxy =
+        new ProxyNodeWebsockets(config -> null, node, /* gridSubPath= */ "");
+
+    Optional<Consumer<Message>> result =
+        proxy.apply("/session/" + sessionId + "/se/cdp", msg -> {});
+
+    assertThat(result).isEmpty();
+    assertThat(acquireCount.get()).isEqualTo(1);
+    // The fix: slot must be released so the counter does not leak.
+    assertThat(releaseCount.get()).isEqualTo(1);
+  }
+
+  @Test
+  void shouldNotReleaseConnectionSlotWhenLimitAlreadyReached() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    // Node reports the connection limit is already reached.
+    Node node =
+        new CountingStubNode(sessionId, null, new AtomicInteger(), releaseCount) {
+          @Override
+          public boolean tryAcquireConnection(SessionId id) {
+            return false;
+          }
+        };
+
+    ProxyNodeWebsockets proxy =
+        new ProxyNodeWebsockets(config -> null, node, /* gridSubPath= */ "");
+
+    Optional<Consumer<Message>> result =
+        proxy.apply("/session/" + sessionId + "/se/cdp", msg -> {});
+
+    assertThat(result).isEmpty();
+    // Nothing was acquired, so nothing should be released.
+    assertThat(releaseCount.get()).isEqualTo(0);
+  }
+
+  @Test
+  void shouldReleaseConnectionSlotWhenGetSessionThrows() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger acquireCount = new AtomicInteger();
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    // getSession throws so no endpoint is ever reached; the slot must still be released.
+    Node node =
+        new CountingStubNode(sessionId, null, acquireCount, releaseCount) {
+          @Override
+          public Session getSession(SessionId id) {
+            throw new NoSuchSessionException("session gone");
+          }
+        };
+
+    ProxyNodeWebsockets proxy =
+        new ProxyNodeWebsockets(config -> null, node, /* gridSubPath= */ "");
+
+    Optional<Consumer<Message>> result =
+        proxy.apply("/session/" + sessionId + "/se/cdp", msg -> {});
+
+    assertThat(result).isEmpty();
+    assertThat(acquireCount.get()).isEqualTo(1);
+    assertThat(releaseCount.get()).isEqualTo(1);
+  }
+
+  @Test
+  void shouldReleaseConnectionSlotWhenBiDiUriIsInvalid() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger acquireCount = new AtomicInteger();
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    // An invalid se:gridWebSocketUrl causes URISyntaxException in findBiDiEndpoint, which
+    // returns Optional.empty() — the slot must still be released.
+    Session session =
+        new Session(
+            sessionId,
+            URI.create("http://localhost:4444"),
+            new ImmutableCapabilities(),
+            new ImmutableCapabilities("se:gridWebSocketUrl", "not a valid uri"),
+            Instant.now());
+
+    Node node = new CountingStubNode(sessionId, session, acquireCount, releaseCount);
+
+    ProxyNodeWebsockets proxy =
+        new ProxyNodeWebsockets(config -> null, node, /* gridSubPath= */ "");
+
+    Optional<Consumer<Message>> result =
+        proxy.apply("/session/" + sessionId + "/se/bidi", msg -> {});
+
+    assertThat(result).isEmpty();
+    assertThat(acquireCount.get()).isEqualTo(1);
+    assertThat(releaseCount.get()).isEqualTo(1);
+  }
+
+  /**
+   * Fix 3: se:vncLocalAddress capability absent from caps. Before the fix, {@code new URI(null)}
+   * threw NullPointerException which was not caught by the URISyntaxException handler, bypassing
+   * the endpoint.isEmpty() release guard and leaking the slot.
+   */
+  @Test
+  void shouldReleaseConnectionSlotWhenVncAddressIsNull() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger acquireCount = new AtomicInteger();
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    // No se:vncLocalAddress in caps → getCapability returns null → was NPE before the fix.
+    Session session =
+        new Session(
+            sessionId,
+            URI.create("http://localhost:4444"),
+            new ImmutableCapabilities(),
+            new ImmutableCapabilities(),
+            Instant.now());
+
+    Node node = new CountingStubNode(sessionId, session, acquireCount, releaseCount);
+
+    ProxyNodeWebsockets proxy =
+        new ProxyNodeWebsockets(config -> null, node, /* gridSubPath= */ "");
+
+    Optional<Consumer<Message>> result =
+        proxy.apply("/session/" + sessionId + "/se/vnc", msg -> {});
+
+    assertThat(result).isEmpty();
+    assertThat(acquireCount.get()).isEqualTo(1);
+    assertThat(releaseCount.get()).isEqualTo(1);
+  }
+
+  @Test
+  void shouldReleaseConnectionSlotWhenVncUriIsInvalid() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger acquireCount = new AtomicInteger();
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    Session session =
+        new Session(
+            sessionId,
+            URI.create("http://localhost:4444"),
+            new ImmutableCapabilities(),
+            new ImmutableCapabilities("se:vncLocalAddress", "not a valid uri %%"),
+            Instant.now());
+
+    Node node = new CountingStubNode(sessionId, session, acquireCount, releaseCount);
+
+    ProxyNodeWebsockets proxy =
+        new ProxyNodeWebsockets(config -> null, node, /* gridSubPath= */ "");
+
+    Optional<Consumer<Message>> result =
+        proxy.apply("/session/" + sessionId + "/se/vnc", msg -> {});
+
+    assertThat(result).isEmpty();
+    assertThat(acquireCount.get()).isEqualTo(1);
+    assertThat(releaseCount.get()).isEqualTo(1);
+  }
+
+  /**
+   * Fix 4: ForwardingListener.onError must release the connection slot in case the WebSocket
+   * implementation does not call onClose after an error. Without the fix the slot leaks permanently
+   * when only onError fires.
+   */
+  @Test
+  void shouldReleaseConnectionSlotWhenWebSocketOnErrorFires() {
+    SessionId sessionId = new SessionId(UUID.randomUUID());
+    AtomicInteger acquireCount = new AtomicInteger();
+    AtomicInteger releaseCount = new AtomicInteger();
+
+    // se:gridWebSocketUrl present and valid so findBiDiEndpoint reaches createWsEndPoint.
+    Session session =
+        new Session(
+            sessionId,
+            URI.create("http://localhost:4444"),
+            new ImmutableCapabilities(),
+            new ImmutableCapabilities("se:gridWebSocketUrl", "ws://localhost:9222/devtools"),
+            Instant.now());
+
+    Node node = new CountingStubNode(sessionId, session, acquireCount, releaseCount);
+
+    // Simulates a WebSocket implementation that fires onError without a subsequent onClose.
+    HttpClient.Factory clientFactory =
+        config ->
+            new HttpClient() {
+              @Override
+              public HttpResponse execute(HttpRequest req) {
+                return new HttpResponse();
+              }
+
+              @Override
+              public WebSocket openSocket(HttpRequest req, WebSocket.Listener listener) {
+                listener.onError(new RuntimeException("network error — no onClose follows"));
+                // Return a no-op WebSocket handle; the connection has already errored.
+                return new WebSocket() {
+                  @Override
+                  public WebSocket send(Message message) {
+                    return this;
+                  }
+
+                  @Override
+                  public void close() {}
+                };
+              }
+
+              @Override
+              public <T>
+                  java.util.concurrent.CompletableFuture<java.net.http.HttpResponse<T>>
+                      sendAsyncNative(
+                          java.net.http.HttpRequest request,
+                          java.net.http.HttpResponse.BodyHandler<T> handler) {
+                throw new UnsupportedOperationException();
+              }
+
+              @Override
+              public <T> java.net.http.HttpResponse<T> sendNative(
+                  java.net.http.HttpRequest request,
+                  java.net.http.HttpResponse.BodyHandler<T> handler) {
+                throw new UnsupportedOperationException();
+              }
+            };
+
+    ProxyNodeWebsockets proxy = new ProxyNodeWebsockets(clientFactory, node, /* gridSubPath= */ "");
+
+    proxy.apply("/session/" + sessionId + "/se/bidi", msg -> {});
+
+    assertThat(acquireCount.get()).isEqualTo(1);
+    // Slot must be released via onError even though onClose was never called.
+    assertThat(releaseCount.get()).isEqualTo(1);
+  }
+
+  /** Minimal Node stub that counts tryAcquireConnection / releaseConnection calls. */
+  private abstract static class StubNode extends Node {
+
+    StubNode(NodeId nodeId, URI uri) {
+      super(DefaultTestTracer.createTracer(), nodeId, uri, SECRET, Duration.ofSeconds(30));
+    }
+
+    @Override
+    public Either<WebDriverException, CreateSessionResponse> newSession(
+        CreateSessionRequest sessionRequest) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public HttpResponse executeWebDriverCommand(HttpRequest req) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Session getSession(SessionId id) throws NoSuchSessionException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public HttpResponse uploadFile(HttpRequest req, SessionId id) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public HttpResponse downloadFile(HttpRequest req, SessionId id) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void stop(SessionId id) throws NoSuchSessionException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isSupporting(Capabilities capabilities) {
+      return false;
+    }
+
+    @Override
+    public NodeStatus getStatus() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public HealthCheck getHealthCheck() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void drain() {
+      draining.set(true);
+    }
+
+    @Override
+    public boolean isReady() {
+      return true;
+    }
+  }
+
+  private static class CountingStubNode extends StubNode {
+
+    private final SessionId ownedSession;
+    private final Session session;
+    private final AtomicInteger acquireCount;
+    private final AtomicInteger releaseCount;
+
+    CountingStubNode(
+        SessionId ownedSession,
+        Session session,
+        AtomicInteger acquireCount,
+        AtomicInteger releaseCount) {
+      super(new NodeId(UUID.randomUUID()), URI.create("http://localhost:5555"));
+      this.ownedSession = ownedSession;
+      this.session = session;
+      this.acquireCount = acquireCount;
+      this.releaseCount = releaseCount;
+    }
+
+    @Override
+    public boolean isSessionOwner(SessionId id) {
+      return ownedSession.equals(id);
+    }
+
+    @Override
+    public boolean tryAcquireConnection(SessionId id) {
+      acquireCount.incrementAndGet();
+      return true;
+    }
+
+    @Override
+    public void releaseConnection(SessionId id) {
+      releaseCount.incrementAndGet();
+    }
+
+    @Override
+    public Session getSession(SessionId id) {
+      return session;
+    }
+  }
+}

